ballerina/dispatcher_service.bal

```diff
@@ -0,0 +1,324 @@
+                return <http:BadRequest>{
+                    body: {
+                        jsonrpc: JSONRPC_VERSION,
+                        'error: {
+                            code: METHOD_NOT_FOUND,
+                            message: "Unknown notification method"
+                        }
+                    }
+                };
+            }
+        }
+    }
+
+    private isolated function handleInitializeRequest(JsonRpcRequest jsonRpcRequest) returns http:BadRequest|http:Ok {
+        JsonRpcRequest {jsonrpc, id, ...request} = jsonRpcRequest;
+        InitializeRequest|error initRequest = request.cloneWithType(InitializeRequest);
+        if initRequest is error {
+            JsonRpcError jsonRpcError = self.createErrorResponse(id, INVALID_REQUEST,
+                string `Invalid request: ${initRequest.message()}`);
+            return <http:BadRequest>{
+                body: jsonRpcError
+            };
+        }
+
+        lock {
+            // If it's a server with session management and the session ID is already set we should reject the request
+            // to avoid re-initialization.
+            if self.isInitialized && self.sessionId != () {
+                JsonRpcError jsonRpcError = self.createErrorResponse(id, INVALID_REQUEST,
+                    "Invalid Request: Only one initialization request is allowed");
+                return <http:BadRequest>{
+                    body: jsonRpcError.cloneReadOnly()
+                };
+            }
+
+            self.isInitialized = true;
+            self.sessionId = uuid:createRandomUuid();
+
+            string requestedVersion = initRequest.params.protocolVersion;
+            string protocolVersion = self.selectProtocolVersion(requestedVersion);
+
+            return <http:Ok>{
+                headers: {
+                    [SESSION_ID_HEADER]: self.sessionId ?: ""
+                },
+                body: {
+                    jsonrpc: JSONRPC_VERSION,
+                    id: id,
+                    result: <InitializeResult>{
+                        protocolVersion: protocolVersion,
+                        capabilities: (self.serverConfigs?.options?.capabilities ?: {}).cloneReadOnly(),
+                        serverInfo: (self.serverConfigs?.serverInfo ?: {
+                            name: "MCP Server",
+                            version: "1.0.0"
+                        }).cloneReadOnly()
+                    }
+                }
+            };
+        }
+    }
+
+    private isolated function handleListToolsRequest(JsonRpcRequest request) returns http:BadRequest|http:Ok {
+        lock {
+            // Check if initialized
+            if !self.isInitialized {
+                JsonRpcError jsonRpcError = self.createErrorResponse(request.id, INVALID_REQUEST,
+                    "Client must be initialized before making requests");
+                return <http:BadRequest>{
+                    body: jsonRpcError.cloneReadOnly()
```

Comment:
Would it be possible to return a readonly record directly from createErrorResponse? This could help eliminate the need for repeated `cloneReadOnly()` calls across the codebase.